### ISSUE TITLE
Numba Sort is Significantly Slower Than NumPy Sort

### ISSUE THREAD

**Issue by seanlaw** [numba/numba#5373]
We are leveraging Numba in the STUMPY package and @mexxexx and I are seeing a significant difference in performance when sorting a NumPy array. The original profiling can be found [here](https://github.com/TDAmeritrade/stumpy/issues/135#issuecomment-596997032) but I'll paste the results below:

> I did some profiling regarding the sorting aspect (jupyter notebook in firefox, Intel(R) Core(TM) i5-6500 CPU @ 3.20GHz, 4 threads):

```
def sort_numpy(T, d, k):
    return np.sort(T, axis=0)

@njit(parallel=True, fastmath=True)
def sort_numba(T, d, k):
    for i in prange(k):
        # T[:,i] = np.sort(T[:, i]) ## no assignment as we don't want to actually sort T
        np.sort(T[:, i])

T = np.random.uniform(-100, 100, (d, k))
```

>I precompiled `sort_numba`. and got the following results:

`d=5`, `k=10000`:
```
>>> %timeit -n 1000 sort_numpy(T, d, k)
295 µs ± 10.1 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
>>> %timeit -n 100 sort_numba(T, d, k)
6.69 ms ± 30.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

`d=5`, `k=100000`:
```
>>> %timeit -n 1000 sort_numpy(T, d, k)
3.08 ms ± 60.8 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
>>> %timeit -n 100 sort_numba(T, d, k)
66.2 ms ± 82.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

`d=50`, `k=10000`:
```
>>> %timeit -n 1000 sort_numpy(T, d, k)
11.2 ms ± 81.4 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
>>> %timeit -n 1000 sort_numba(T, d, k)
6.58 ms ± 10.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

> So sorting with numpy is orders of magnitude faster for a small number of dimensions and long time serieses, the case that we usually have.

Is this expected or are we doing something wrong here?

**Comment by mexxexx** [comment]
Also as a further question: Since numba does not support sorting using the `axis` keyword in nopython mode yet, is it somehow possible to split up the sorting into chunks that do the numpy sorting in parallel? (In our case `k >> d`, where d generally is less that 50, but `k` can be in the range of multiple millions.

**Comment by stuartarchibald** [comment]
Thanks for the report. There's a note here: http://numba.pydata.org/numba-doc/latest/reference/numpysupported.html#other-methods

> Sorting may be slightly slower than Numpy’s implementation.

So I believe what you see may be expected.

**Comment by seanlaw** [comment]
@stuartarchibald We saw that note but perhaps our interpretation of "slightly slower" is different. I take that to mean, at worst, 20% slower. For a length an array length of `100000`, we're seeing an order of magnitude difference which is significantly more than "slightly slower". 

> Since numba does not support sorting using the axis keyword in nopython mode yet, is it somehow possible to split up the sorting into chunks that do the numpy sorting in parallel? (In our case k >> d, where d generally is less that 50, but k can be in the range of multiple millions.

Any chance you can shed some light on this as well?

**Comment by stuartarchibald** [comment]
I've taken a look at this. I'm wondering if this is perhaps down to memory churn? By operating on a slice there's an allocation somewhere as the sort is done out-of-place.

**Comment by stuartarchibald** [comment]
> Any chance you can shed some light on this as well? We love Numba so any suggestions would be great appreciated!

Is `prange` not working for this?

**Comment by seanlaw** [comment]
Yes, the timing shown above is with `prange`. `prange` is "working" but it is still an order of magnitude slower than NumPy

**Comment by seanlaw** [comment]
> I've taken a look at this. I'm wondering if this is perhaps down to memory churn? By operating on a slice there's an allocation somewhere as the sort is done out-of-place.

I don't think I know enough about Numba internals to be able to understand what is happening under the hood. Are you able to reproduce the timing comparison(s) above?

**Comment by stuartarchibald** [comment]
@hameerabbasi is hoping to work on this for 0.53, xref: https://github.com/numba/numba/wiki/Minutes_2020_10_20

**Comment by seanlaw** [comment]
@hameerabbasi Were you able to make progress on this? Anything we can do to help?

**Comment by hameerabbasi** [comment]
I can look into it, but the difference certainly shouldn't be an order of magnitude. My proposal was to replace merge sort by radix sort like in NumPy, which can be faster in some cases but not all. It shouldn't have an effect on your case as you don't use merge sort (it's only used if you pass a kwarg).

So I'm not entirely sure what is happening here.

**Comment by seanlaw** [comment]
@hameerabbasi I think that a good place to start is to see if you are able to reproduce the tests at the top of this thread. Would you mind confirming?

**Comment by hameerabbasi** [comment]
TL;DR I can reproduce the results

Script: [tmp.py.zip](https://github.com/numba/numba/files/8692744/tmp.py.zip)
Results:

```
The mean time for sort_numpy was 493.91984939575195 μs and the standard deviation was 11.842128855966518 μs.
OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.
The mean time for sort_numba was 577.0373344421387 μs and the standard deviation was 92.17707428796716 μs.
The mean time for sort_numpy was 5011.153221130371 μs and the standard deviation was 135.17833508354994 μs.
The mean time for sort_numba was 7179.431915283203 μs and the standard deviation was 421.8190407178981 μs.
The mean time for sort_numpy was 10645.644664764404 μs and the standard deviation was 242.4166365658219 μs.
The mean time for sort_numba was 5054.254531860352 μs and the standard deviation was 707.8448591388241 μs.
```

This is on an 8-core MacBook Pro 14", 2021.

**Comment by esc** [comment]
I can repro this too:

```
 💣 zsh» python tmp.py
The mean time for sort_numpy was 488.8987541198731 μs and the standard deviation was 43.920950070242135 μs.
OMP: Info #271: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.
The mean time for sort_numba was 652.3394584655762 μs and the standard deviation was 44.03392231405237 μs.
The mean time for sort_numpy was 5185.270309448242 μs and the standard deviation was 628.184829390657 μs.
The mean time for sort_numba was 7743.899822235107 μs and the standard deviation was 457.2864641283301 μs.
The mean time for sort_numpy was 9848.854541778564 μs and the standard deviation was 871.6013331967821 μs.
The mean time for sort_numba was 4342.429637908936 μs and the standard deviation was 781.7597993442256 μs.
python tmp.py  24.66s user 0.57s system 382% cpu 6.601 total
```

Machine details:

<img width="264" alt="Screen Shot 2022-05-16 at 08 26 51" src="https://user-images.githubusercontent.com/41563/168532384-59f7d555-0aed-4de8-b7ce-dc9f9956eb96.png">

**Comment by mariusmuja** [comment]
Also seeing a significant performance difference between numpy and numba sort:
```python
@jit(nopython=True, fastmath=True)
def numba_sort(b):
    return np.sort(b)
```
Testing with a large int array:
```python
In [6]: b.dtype
Out[6]: dtype('uint32')

In [7]: len(b)
Out[7]: 82843955
```

Sorting times:

```python
In [12]: %timeit np.sort(b[:100000])
943 µs ± 9.68 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [13]: %timeit np.sort(b[:1000000])
9.1 ms ± 220 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [19]: %timeit np.sort(b[:10000000])
123 ms ± 1.57 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [14]: %timeit numba_sort(b[:100000])
76.2 ms ± 491 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [15]: %timeit numba_sort(b[:1000000])
4.6 s ± 37.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [20]: %timeit numba_sort(b[:10000000])
1min 3s ± 81.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

**Comment by mariusmuja** [comment]
Here's an easy to reproduce example:

```python
In [2]: a = np.arange(100000)

In [4]: %timeit np.sort(a)
485 µs ± 8.15 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [8]: %timeit numba_sort(a)
754 µs ± 27.5 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [10]: a[len(a)//2] = 100

In [11]: %timeit np.sort(a)
680 µs ± 8.97 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [12]: %timeit numba_sort(a)
286 ms ± 3.29 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

**Comment by hameerabbasi** [comment]
What this tells me from my knowledge of sorting is that Numba is doing a quick sort instead of an introsort, which is asymptotically worse for almost sorted arrays.

**Comment by lesshaste** [comment]
Were there any thoughts about addressing this? There are well documented sorting algorithms that are much faster when the input is nearly sorted .

**Comment by esc** [comment]
> Were there any thoughts about addressing this? There are well documented sorting algorithms that are much faster when the input is nearly sorted .

No, fixing sorting isn't on any TODO list that I am aware of. Pull-requests welcome of course.

**Comment by seanlaw** [comment]
> @hameerabbasi is hoping to work on this for 0.53, xref: https://github.com/numba/numba/wiki/Minutes_2020_10_20

@esc @hameerabbasi As always, thank you for your time! Did this ever become prioritized for 0.53 as a part of the aforementioned minutes?

**Comment by hameerabbasi** [comment]
Hey @seanlaw , I implemented Radix sort, but unfortunately the performance was slower than existing algorithms, so I didn't make a PR.

**Comment by lesshaste** [comment]
Can you show your radix sort code and timings?

**Comment by ghuls** [comment]
Codon now has a numpy implementation: https://www.exaloop.io/blog/codon-2025

They implemented multiple sorting algorithms for sorting numpy arrays:
https://github.com/exaloop/codon/blob/develop/stdlib/numpy/sorting.codon

```python
import numpy as np
import numba
import codon

@numba.jit
def sort_array_numba(x):
    return np.sort(x)

@codon.jit
def sort_array_codon(x):
    import numpy as np
    return np.sort(x)

@codon.jit
    def func_codon(x):
    return 0.0


In [24]: x = np.random.rand(100000)

In [25]: %timeit y = np.sort(x)
4.71 ms ± 21.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [26]: %timeit y = sort_array_numba(x)
5.67 ms ± 26.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [27]: %timeit y = sort_array_codon(x)
1.1 ms ± 6.3 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [28]: X = np.random.rand(10000000)

In [29]: %timeit Y = np.sort(X)
670 ms ± 6.16 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [30]: %timeit Y = sort_array_numba(X)
803 ms ± 1.08 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [31]: %timeit Y = sort_array_codon(X)
134 ms ± 747 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [32]: xs = np.random.rand(100)

In [33]: %timeit ys = np.sort(xs)
1.17 µs ± 6.7 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [34]: %timeit ys = sort_array_numba(xs)
1.53 µs ± 8.26 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [35]: %timeit ys = sort_array_codon(xs)
26.3 µs ± 120 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


# Codon-jit python interaction overhead is relatively big for small functions:
In [41]: %timeit ys = func_codon(xs)
22.2 µs ± 106 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
